### PR TITLE
[WIP] cmake 3.18 upgrade w/ hunter package management

### PR DESCRIPTION
**Pull request by millerhooks**
In an attempt to wrangle `RAPIDS` `CMakeLists.txt` I've adopted the use of `hunter` as a package manager and audited some other issues. This PR is not meant to dictate where to go with `cmake` but just an attempt to solve a few problems and get some discussion going about solving problems going forward. Just to try to streamline that discussion I've outlined what I'm doing and why below as well as some excerpts from the `hunter` documentation. 

My tear down and rebuild process for `cudf` was all in the context of `rapids-compose` so I could easily destroy `ccache` and `conda` to verify multiple states. I do not think this PR fully reflects the MVP to merge this pattern. Also in that same spirit, this PR likely breaks the build process when using `make` to start `rapids-compose` as it bypasses `conda` dependencies in favor of building from scratch from source for a few things (`boost`). Doing a `make` that fails because of `boost` linking is normal at this phase. After that building from inside the container via `make rapids.run` inside `~/cudf` via `build.sh` should produce the correct results. Details of the how and why of this are outlined below.

Problems and Goals
==================

Problems
--------
* Our `cmake` (in all `RAPIDS` projects) has accumulated some neglected copypasta that bogs down production and steals productivity 
* We cannot upgrade to the newest `cmake` (3.18) which has very useful helpers for `CUDA` support because of issues relating to our complicated environment flow
* Not having a simple, standards driven `cmake` causes repetition of bad but functional tooling
* `cmake` is an afterthought for many developers slowing down production when `cmake` ontology is not clear
* New `cmake 3.18` `CUDA` features are broken when we do our `conda` environment steps in our `CMakeList.txt`
  - Do we still need our `CUDA` architecture detection?
  - In some environments [`rapids-compose`] does our current `conda` environment break the new (`cmake 3.18`) `nvcc` / `cuda` detection?
    * `/conda/cuda_10.2/envs/rapids/compiler_compat/ld: cannot find -lcudf`

Goals
-----
* Make `RAPIDS` `cmake` compatible with `3v.18` to utilize new native `CUDA` features
* Do not build a new `cmake` package management paradigm. 
  - `hunter` is pretty cool. `CPM` is alright. I picked `hunter`
* Decrease `RAPIDS` install conflicts related to existing system installed packages
* Pair down redundant and complicated environment vars, includes, and linked directories
* Speed up build times
* Increase resilliance and predictability to dependency management
* A single source of truth for future configurations in `RAPIDS` projects
* Easy incremental integration without disrupting development
* Easy CI/CD integration
* Easy improvements, enhancements, and ability to refactor for `cmake` configuration for all rapids projects
* Reduce malignant duplication between `RAPIDS` projects `cmake` configurations

Overview of Changes
===================
* Adding `hunter` as the `cmake` package manager.
  - https://github.com/cpp-pm/hunter
  - https://hunter.readthedocs.io/en/latest/
* Simplifying `zlib`, `boost-filesystem`, `gtest`, and `benchmarks` management via `hunter`
* [WIP] Using new `FindCUDAToolkit` features of `cmake 3.18`.
  - Currently the detection of the package is failing, though it seems to be able to identify the installed `CUDA` version.
* [TODO] Build `arrow` for `hunter` to simplify the `arrow` dependency install process

Testing Procedure
=================
To properly test this you should use Paul Taylor's [@trxcllnt] `rapids-compose` project [github.com/trxcllnt/rapids-compose] and not your system directly. This will make it easier to modify environment to understand all the variable involved and also not waste too much time and possibly upset your normal development environment.

Some Notes About `ccache`
-------------------------
There are lots of places to keep binaries from compile to speed things up. `hunter` takes the philosiphy that unless you specify to build without `hunter` using the flag `-DHUNTER_ENABLED=NO` you are opting into a system that builds from source and does not search the system for libraries. When `-DHUNTER_ENABLED=NO` is set, cmake should operate as normal using local dependencies. 

https://hunter.readthedocs.io/en/latest/overview/compatibility.html

The upshot is that by default you are ignoring system installs to avoid conflict. This means slow build times for outside or development users unless you specify a `ccache` repository for hunter to pull from. The benefit to that is that because `hunter` offers many simple ways to create that `ccache` repository you get versioned isolation of binaries. One of those options is just using a git repository. Making it very easy to isolate dependencies by version and provide support for custom builds (e.g. Arrow).

I realize that this is a big consideration but I think the value of the paradigm speaks for itself. However, I do not want to assume I know where we think this project should go or what an adoption of `hunter` means for people that work on and customers that use `RAPIDS`. I think it's important that appropriate stakeholders weigh in on this. For this PoC I have left the cache setup to the default of `~/.hunter` to avoid that rabit hole without knowing whats best. Because of that when you build this the first time you will experience some delays in building that are normally quick steps because of referencing host packages like `zlib` and `boost`.

I also did not configure `hunter` to use `~/.ccache` which would be in line with how Paul Taylor's [@trxcllnt] `rapids-compose` project [github.com/trxcllnt/rapids-compose] works. They can very much piggy back off of each other, but after a bit of assesemnt I didn't want to define something that would work against the dependency versioning features of `hunter`.

Using GitHub repository as binary cache server
----------------------------------------------

It is possible to upload Hunter binary cache to the server. Next shown an example of using GitHub as a hosting. All big raw *.tar.bz2 archives uploaded as assets to release with names cache-* (directory layout does not matter) and all small text files with meta information uploaded directly to branch master (directory layout matters) (see hunter-cache as example).

https://hunter.readthedocs.io/en/latest/user-guides/hunter-user/github-cache-server.html#uploading-to-server

More About Hunter
=================

I am suggesting some large changes and a new dependency (`hunter`). Below I'll pull a few things from the `hunter` docs
to give an overview of the reasons I like `hunter` and more easily fascillitate discussion.
 
**Cmake Hunter Environment**
Cmake Hunter Environment

https://hunter.readthedocs.io/en/latest/user-guides/regular.html

Don't Repeat Yourself
---------------------

```
If you are already familiar with ExternalProject_Add and have written some super-build projects before, you are probably already aware that writing a complete solution with toolchains, build types, build options, parallel jobs, forwarding of compiler flags, and making it work correctly for all generators is not a trivial task.

Hunter stores ExternalProject_Add recipes as a set of templates. Once written, formula (build scheme) can be reused by other projects without copying the collection of super-build files (DRY principle). When a new package with a new scheme is introduced, all you need to do is just update the SHA1/URL of HunterGate command.
```

https://hunter.readthedocs.io/en/latest/overview/dry.html

Compared to a ‘requirements.txt’ style approach
-----------------------------------------------

Note that Hunter’s approach differs from a requirements.txt-like approach (i.e. when external packages are specified in a separate file). This allows Hunter to avoid duplication of logic in many cases, even if the requirements.txt style approach also automatically downloads dependencies too.

Imagine that we have to specify dependencies in some kind of requirements.cmake file and there is a user option BUILD_WITH_BAR:

```
# requirements.cmake

if(WIN32 AND BUILD_WITH_BAR)
  command_to_install(Bar)
endif()
```

Or, in the case that it isn’t CMake code, this might by something fancy like requirements.json:

```
{
  "dependencies":
  {
    "package": "Bar",
    "platform": "windows",
    "cmake option": "BUILD_WITH_BAR"
  }
}
```
You would have to repeat the same condition in the CMakeLists.txt file:

```
# requirements.cmake

if(WIN32 AND BUILD_WITH_BAR)
  command_to_install(Bar)
endif()
# CMakeLists.txt

if(WIN32 AND BUILD_WITH_BAR)
  find_package(Bar CONFIG REQUIRED)
  target_compile_definitions(... PUBLIC "WITH_BAR")
endif()
```

Later, when you need to change this dependency in CMakeLists.txt, you’d better not forget to also modify requirements.cmake accordingly. Remember real world libraries can have nontrivial chain of conditions, e.g. OpenCV components.

https://hunter.readthedocs.io/en/latest/overview/automatic.html#compared-to-a-requirements-txt-style-approach

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by harrism** [comment]
Definitely not 0.15.

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -239,39 +248,18 @@ file(INSTALL ${LIBCXX_INCLUDE_DIR}     DESTINATION ${CMAKE_BINARY_DIR}/include/l
...
-    message(FATAL_ERROR "ZLib not found, please check your settings.")
-endif(ZLIB_FOUND)
+hunter_add_package(ZLIB)
+find_package(ZLIB CONFIG REQUIRED)
 
 ###################################################################################################
 # - find boost ------------------------------------------------------------------------------------
 
 # Don't look for a CMake configuration file
 set(Boost_NO_BOOST_CMAKE ON)
```
Comment (kkraus14) [inline cpp/CMakeLists.txt:258]: Can we remove this now?

**Comment by harrism** [comment]
There is also https://conan.io/

**Comment by kkraus14** [comment]
> There is also https://conan.io/

There's also `vcpkg`: https://vcpkg.readthedocs.io/en/latest/

vcpkg and conan are more dedicated to downloading binaries, which I don't think we want to do. Hunter allows for building from source but it looks non-trivial to control how things are built from source / adding a non-hunter enabled package. `CPM` (https://github.com/TheLartians/CPM.cmake) handles finding an existing package and building from source.

**Comment by harrism** [comment]
Can we close this?

**Comment by harrism** [comment]
Ping @millerhooks @kkraus14 moving to 0.17.

**Comment by kkraus14** [comment]
@millerhooks can this PR just be closed? I believe we've moved in a separate direction using CPM.

**Comment by harrism** [comment]
@kkraus14 can this be closed?

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by kkraus14** [comment]
Yes, closing.

**Pull request closed without merging**